src/ucp/core/ucp_mm.c

```diff
@@ -1305,3 +1386,226 @@ ucs_status_t ucp_mem_reg_md_map_update(ucp_context_h context)
+        memh->md_map       |= UCS_BIT(md_index);
+
+        ucs_trace("imported address %p length %zu on md[%d]=%s: uct_memh %p",
+                  ucp_memh_address(memh), ucp_memh_length(memh), md_index,
+                  context->tl_mds[md_index].rsc.md_name,
+                  memh->uct[md_index]);
+    }
+
+    if (memh->md_map == 0) {
+        ucs_error("no suitable UCT memory domains to perform importing on");
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_memh_import_slow(ucp_context_h context, ucs_rcache_t *existing_rcache,
+                     ucp_mem_h user_memh,
+                     ucp_unpacked_exported_memh_t *unpacked)
+{
+    ucs_rcache_t *rcache;
+    ucs_status_t status;
+    ucp_mem_h memh;
+    char rcache_name[128];
+    khiter_t iter;
+    int ret;
+
+    ucs_assert(user_memh != NULL);
+
+    UCP_THREAD_CS_ENTER(&context->mt_lock);
+
+    if (context->imported_mem_hash != NULL) {
+        if (existing_rcache == NULL) {
+            ucs_snprintf_safe(rcache_name, sizeof(rcache_name),
+                              "ucp_import_rcache[0x%" PRIx64 "]",
+                              unpacked->remote_uuid);
+            status = ucp_mem_rcache_create(context, rcache_name, &rcache);
+            if (status != UCS_OK) {
+                goto out;
+            }
+
+            iter = kh_put(ucp_context_imported_mem_hash,
+                          context->imported_mem_hash, unpacked->remote_uuid,
+                          &ret);
+            ucs_assertv((ret != UCS_KH_PUT_FAILED) &&
+                        (ret != UCS_KH_PUT_KEY_PRESENT), "ret %d", ret);
+
+            kh_value(context->imported_mem_hash, iter) = rcache;
+        } else {
+            rcache = existing_rcache;
+        }
+
+        status = ucp_memh_rcache_get(rcache, unpacked->address,
+                                     unpacked->length, unpacked->mem_type,
+                                     &memh);
+        if (status != UCS_OK) {
+            goto err_rcache_destroy;
+        }
+
+        memh->flags       |= UCP_MEMH_FLAG_IMPORTED;
+        user_memh->parent  = memh;
+    } else {
+        memh   = user_memh;
+        rcache = NULL;
+    }
+
+    memh->reg_id      = unpacked->reg_id;
+    memh->remote_uuid = unpacked->remote_uuid;
+    status            = ucp_memh_import_attach(context, memh,
+                                               unpacked->tl_mkeys,
+                                               unpacked->num_tl_mkeys);
+    if (status != UCS_OK) {
+        goto err_memh_free;
+    }
+
+out:
+    UCP_THREAD_CS_EXIT(&context->mt_lock);
+    return status;
+
+err_memh_free:
+    if (rcache != NULL) {
+        ucs_rcache_region_put_unsafe(rcache, &memh->super);
+    }
+err_rcache_destroy:
+    if ((rcache != NULL) && (rcache != existing_rcache)) { 
+        /* Rcache was allocated here - remove it from hash and destroy */
+        iter = kh_get(ucp_context_imported_mem_hash,
```

Comment:
when can be iter invalid? we are here only when we created rcache in this func and iter has to be already valid

Reply:
it is a protection if `ucp_context_imported_mem_hash` was changed and `iter` will be invalidated, but currently it is not a case.
also, Coverity issue can be raised, because it may not understand that `iter` is initialized.